### PR TITLE
TF-3034 Prevent duplicate draft warning

### PR DESCRIPTION
**Pull request by tddang-linagora**
## Issue
- #3034 

## Demo
- Compose new email

https://github.com/user-attachments/assets/8f1fb7c8-f63e-4cb1-b432-ae07933bdb4e

- Edting draft email

https://github.com/user-attachments/assets/76cb7626-bf22-41d0-956e-f132840802fd

- Mobile (both cases)
[mobile_draft_save.webm](https://github.com/user-attachments/assets/ccf24d66-54b5-4872-b395-a11897d92238)

### CODE REVIEW

**Review by dab246 — COMMENTED** [review]
- [ ] Add demo for mobile

File: test/features/composer/presentation/composer_controller_test.dart [hunk 1]
```diff
@@ -0,0 +1,852 @@
...
+
+              return PermissionStatus.denied.index;
+            }
+          );
+      }
+
+      test(
+        'should update _savedEmailDraftHash '
+        'when there is a new view state '
+        'and the state is GetAlwaysReadReceiptSettingSuccess',
```
Comment (dab246) [inline test/features/composer/presentation/composer_controller_test.dart:296]: What is the point of this unit test? Why update _savedEmailDraftHash when state is GetAlwaysReadReceiptSettingSuccess

Comment (tddang-linagora) [reply test/features/composer/presentation/composer_controller_test.dart:296]: read receipt is one of the value in the hash, and its value is retrieved from the server. Hence, `GetAlwaysReadReceiptSettingSuccess` is needed.

**Comment by dab246** [comment]
- [ ] Do not show dialog save as draft when compose from `email address` or `mailto` address bar



https://github.com/user-attachments/assets/982950e8-0045-4f46-a1a3-d7024e0df4b1

**Comment by dab246** [comment]
- [ ] Open draft email -> click show full screen, minimize   -> click close -> auto show dialog save as draft


https://github.com/user-attachments/assets/6e0f0eb0-cd52-41c3-a92a-1b5e94f9b251

**Comment by dab246** [comment]
- [ ] Open draft email -> change subject -> reload page ->  click close composer -> do not show dialog save as draft


https://github.com/user-attachments/assets/2d7881cd-1807-498f-9f1f-06eecf8c8691

**Comment by tddang-linagora** [comment]
>Open draft email -> change subject -> reload page -> click close composer -> do not show dialog save as draft

Reload and Session expired cases are ignored on purpose.

**Comment by tddang-linagora** [comment]
>Do not show dialog save as draft when compose from email address or mailto address bar

Emails are automatically filled as expected. That doesn't count as email content changed, just as when you open a draft. I don't see a problem with this.

File: test/features/composer/presentation/composer_controller_test.dart [hunk 2]
```diff
@@ -0,0 +1,852 @@
...
+        final selectedIdentity = Identity(id: IdentityId(Id('alice')));
+        when(mockMailboxDashBoardController.composerArguments).thenReturn(
+          ComposerArguments(
+            selectedIdentityId: selectedIdentity.id,
+            identities: [selectedIdentity]));
+
+        final attachment = Attachment();
+        when(mockUploadController.attachmentsUploaded).thenReturn([attachment]);
+
+        grandPermission(Permission.contacts);
```
Comment (hoangdat) [inline test/features/composer/presentation/composer_controller_test.dart:429]: why this test case run in web, we still need to grand permission?

File: lib/features/composer/presentation/composer_controller.dart [hunk 3]
```diff
@@ -1439,6 +1432,8 @@ class ComposerController extends BaseController with DragDropFileMixin implement
     final selectedIdentityFromHeader = _selectIdentityFromId(identityIdFromHeader);
     if (selectedIdentityFromHeader == null) return;
     identitySelected.value = selectedIdentityFromHeader;
+
+    _updateSavedEmailDraftHash();
```
Comment (hoangdat) [inline lib/features/composer/presentation/composer_controller.dart:1436]: I still not find the test for this case

Comment (hoangdat) [reply lib/features/composer/presentation/composer_controller.dart:1436]: ahh, found it. this case
```
      test(
        'should update _savedEmailDraftHash '
        'when there is a new view state '
        'and the state is GetEmailContentSuccess '
        'and emailActionType is editDraft',
      () async {
```

File: test/features/composer/presentation/composer_controller_test.dart [hunk 4]
```diff
@@ -0,0 +1,852 @@
...
+          PlatformInfo.isTestingForWeb = false;
+        });
+      });
+
+      test(
+        'should update _savedEmailDraftHash '
+        'when there is a new view state '
+        'and the state is GetEmailContentSuccess '
+        'and emailActionType is editDraft',
+      () async {
```
Comment (hoangdat) [inline test/features/composer/presentation/composer_controller_test.dart:796]: I found we call `_savedEmailDraftHash` in `_setUpRequestReadReceiptForDraftEmail` and `_restoreIdentityFromHeader`, can we need to make sure 2 calls of this still not make the hash changed?

**Comment by tddang-linagora** [comment]
>Open draft email -> click show full screen, minimize -> click close -> auto show dialog save as draft

Will be fixed when https://github.com/linagora/tmail-flutter/pull/3191 is merged

**Comment by tddang-linagora** [comment]
>Do not show dialog save as draft when compose from email address or mailto address bar


https://github.com/user-attachments/assets/6fb1045a-9dd8-40dc-8d9b-899105218e8d

**Comment by tddang-linagora** [comment]
>Open draft email -> change subject -> reload page -> click close composer -> do not show dialog save as draft

https://github.com/user-attachments/assets/ce470f7d-212a-4d74-a1aa-62826be18234

File: test/features/composer/presentation/composer_controller_test.dart (hunk 2 above)
Comment (tddang-linagora) [reply test/features/composer/presentation/composer_controller_test.dart]: Permission condition removed

File: test/features/composer/presentation/composer_controller_test.dart (hunk 4 above)
Comment (tddang-linagora) [reply test/features/composer/presentation/composer_controller_test.dart]: Test updated to verify that called twice without changing the hash.

File: lib/features/composer/presentation/composer_controller.dart [hunk 5]
```diff
@@ -1265,13 +1265,18 @@ class ComposerController extends BaseController with DragDropFileMixin implement
...
-    if (draftSaved
-      || composerArguments.value?.emailActionType == EmailActionType.compose
-      || composerArguments.value?.emailActionType == EmailActionType.editDraft
+  Future<void> _updateSavedEmailDraftHash() async {
+    _savedEmailDraftHash = await _hashDraftEmail();
+  }
+
+  Future<void> _initEmailDraftHash() async {
+    if (composerArguments.value?.emailActionType != EmailActionType.compose
+      && composerArguments.value?.emailActionType != EmailActionType.editDraft
```
Comment (hoangdat) [inline lib/features/composer/presentation/composer_controller.dart:1274]: in this case, when we reply/forward/reply_all still not calculate hash

**Comment by dab246** [comment]
> > Open draft email -> click show full screen, minimize -> click close -> auto show dialog save as draft
> 
> Will be fixed when #3191 is merged

Please add demo for it.

**Comment by dab246** [comment]
- [ ] Confirm save as draft always display when click `save as draft` button in bottom bar composer


https://github.com/user-attachments/assets/ceff2566-adfb-4a9d-907e-ab233946a38c

**Comment by dab246** [comment]
- [ ] Confirm save as draft always display when open draft email -> click close button


https://github.com/user-attachments/assets/16a8601c-b3b8-48a7-b015-fcc79f91861e

**Comment by tddang-linagora** [comment]
> > > Open draft email -> click show full screen, minimize -> click close -> auto show dialog save as draft
> > 
> > 
> > Will be fixed when #3191 is merged
> 
> Please add demo for it.


https://github.com/user-attachments/assets/d6a937cb-6eab-41b9-85d6-7b6de7c5ffef

**Comment by dab246** [comment]
- [ ] Cannot show warning dialog save when: open draft mail -> drag `To` to `Cc` -> Click close


https://github.com/user-attachments/assets/7e3a9398-2ec9-4ad2-9d8b-98069e7888e3

**Comment by dab246** [comment]
- [ ] Warning dialog save always display when: open new composer -> enable read recipient -> click close -> save as draft -> open this draft mail -> click close

https://github.com/user-attachments/assets/827b1822-3ce1-4a9d-bb32-a6434a00a552

**Comment by tddang-linagora** [comment]
>Cannot show warning dialog save when: open draft mail -> drag To to Cc -> Click close

https://github.com/user-attachments/assets/527a74d2-449c-4a01-9c47-00302e396033

>Warning dialog save always display when: open new composer -> enable read recipient -> click close -> save as draft -> open this draft mail -> click close

https://github.com/user-attachments/assets/f8984c59-0e92-49c8-94a1-64c4fa702cc8

File: lib/features/composer/presentation/composer_controller.dart [hunk 6]
```diff
@@ -200,6 +200,7 @@ class ComposerController extends BaseController with DragDropFileMixin implement
   ButtonState _sendButtonState = ButtonState.enabled;
   SignatureStatus _identityContentOnOpenPolicy = SignatureStatus.editedAvailable;
   int? _savedEmailDraftHash;
+  bool restoringSignatureButton = false;
```
Comment (dab246) [inline lib/features/composer/presentation/composer_controller.dart:203]: use private variable

File: lib/features/composer/presentation/model/saved_email_draft.dart [hunk 7]
```diff
@@ -28,9 +28,10 @@ class SavedEmailDraft with EquatableMixin {
   List<Object?> get props => [
     content,
     subject,
-    toRecipients,
-    ccRecipients,
-    bccRecipients,
+    // Prevent identical Set<EmailAddress>
+    {0: toRecipients},
+    {1: ccRecipients},
+    {2: bccRecipients},
```
Comment (dab246) [inline lib/features/composer/presentation/model/saved_email_draft.dart:34]: what is this?

Comment (tddang-linagora) [reply lib/features/composer/presentation/model/saved_email_draft.dart:34]: Currently `SavedEmailDraft` `props` is a list, in which there are 3 `Set<EmailAddress` `toRecipients`, `ccRecipients` and `bccRecipients`. When dragging any email from one of the `To/CC/BCC` field to another, as long as the whole `props` doesn't change, the hashcodes are the same. By tagging each of those `Set` with an index, this prevent the bug where the code cannot detect changes when an email is dragged across fields.

File: lib/features/composer/presentation/composer_controller.dart (hunk 6 above)
Comment (tddang-linagora) [reply lib/features/composer/presentation/composer_controller.dart]: Done

File: lib/features/composer/presentation/model/saved_email_draft.dart (hunk 7 above)
Comment (dab246) [reply lib/features/composer/presentation/model/saved_email_draft.dart:34]: We should write unit test for for it

Comment (tddang-linagora) [reply lib/features/composer/presentation/model/saved_email_draft.dart:34]: Done

File: test/features/composer/presentation/model/saved_email_draft_test.dart [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+        hasReadReceipt: false
+      );
+      
+      // act
+      final props = savedEmailDraft.props;
+      
+      // assert
+      expect(props[2], equals({0: savedEmailDraft.toRecipients}));
+      expect(props[3], equals({1: savedEmailDraft.ccRecipients}));
+      expect(props[4], equals({2: savedEmailDraft.bccRecipients}));
```
Comment (dab246) [inline test/features/composer/presentation/model/saved_email_draft_test.dart:28]: Should add unit test to check if `hashCode` is different when having same email address in `to`, `cc`, `bcc`

Comment (tddang-linagora) [reply test/features/composer/presentation/model/saved_email_draft_test.dart:30]: Done

**Comment by dab246** [comment]
- [ ] Please rebase it

**Review by dab246 — APPROVED** [review]

File: test/features/composer/presentation/model/saved_email_draft_test.dart
Comment (hoangdat) [inline test/features/composer/presentation/model/saved_email_draft_test.dart:1]: add 2 more cases:
```
test(
        'when toRecipients updated, hashcode should be changed', () {

          final listToRecipients = {
            EmailAddress('to name', 'to email')
          };
          // arrange
          final savedEmailDraft = SavedEmailDraft(
              subject: 'subject',
              content: 'content',
              toRecipients: listToRecipients,
              ccRecipients: {EmailAddress('cc name', 'cc email')},
              bccRecipients: {EmailAddress('bcc name', 'bcc email')},
              identity: null,
              attachments: [],
              hasReadReceipt: false
          );
          final hashCodeBeforeChange = savedEmailDraft.hashCode;

          listToRecipients.add(EmailAddress('to name 2', 'to email 2'));

          final hashCodeAfterChange = savedEmailDraft.hashCode;

          expect(hashCodeBeforeChange, isNot(hashCodeAfterChange));
        });

    test(
        '2 instances of SavedEmailDraftTest should have same hashcode if all properties are the same', () {

      // arrange
      final savedEmailDraft = SavedEmailDraft(
          subject: 'subject',
          content: 'content',
          toRecipients: {EmailAddress('to name', 'to email')},
          ccRecipients: {EmailAddress('cc name', 'cc email')},
          bccRecipients: {EmailAddress('bcc name', 'bcc email')},
          identity: null,
          attachments: [],
          hasReadReceipt: false
      );

      final savedEmailDraft2 = SavedEmailDraft(
          subject: 'subject',
          content: 'content',
          toRecipients: {EmailAddress('to name', 'to email')},
          ccRecipients: {EmailAddress('cc name', 'cc email')},
          bccRecipients: {EmailAddress('bcc name', 'bcc email')},
          identity: null,
          attachments: [],
          hasReadReceipt: false
      );

      expect(savedEmailDraft.hashCode, equals(savedEmailDraft2.hashCode));
    });
```

Comment (tddang-linagora) [reply test/features/composer/presentation/model/saved_email_draft_test.dart:1]: Done

**Review by hoangdat — APPROVED** [review]

**Comment by dab246** [comment]
- [ ] Please check again warning on CI 

Screenshot 2024-10-04 at 17 45 07

**Review by hoangdat — APPROVED** [review]

**Pull request merged**